Remove wrong internal assert in handle_view_on_rebase

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85849
* #79056
* #88216
* __->__ #88243

Fixes: https://github.com/pytorch/pytorch/issues/88205

The `CreationMeta::NO_GRAD_MODE` path in handle_view_on_rebase wrongly assumes that the tensor would be a leaf, because tensors created in no_grad are always leaf tensors. However, due to creation_meta propagation, a view of a view created in no_grad also has `CreationMeta::NO_GRAD_MODE`, but DOES have grad_fn.

soulitzer committed:
Remove wrong internal assert in handle_view_on_rebase

[ghstack-poisoned]

soulitzer committed:
Update on "Remove wrong internal assert in handle_view_on_rebase"




[ghstack-poisoned]

soulitzer committed:
Update on "Remove wrong internal assert in handle_view_on_rebase"


Fixes: https://github.com/pytorch/pytorch/issues/88205

The `CreationMeta::NO_GRAD_MODE` path in handle_view_on_rebase wrongly assumes that the tensor would be a leaf, because tensors created in no_grad are always leaf tensors. However, due to creation_meta propagation, a view of a view created in no_grad also has `CreationMeta::NO_GRAD_MODE`, but DOES have grad_fn.

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -5881,6 +5881,15 @@ def run_tests(fn):
+        x = torch.zeros(2, requires_grad=False)
+        with torch.no_grad():
+            y = x.view(2)
+            y.requires_grad_(True)
```
Does this line need to be inside the no_grad block?

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -789,7 +789,6 @@ void handle_view_on_rebase(
           " allow the output views to be modified inplace. You should replace the inplace operation by an"
           " out-of-place one.");
     } else if (creation_meta == CreationMeta::NO_GRAD_MODE) {
-      TORCH_INTERNAL_ASSERT(!grad_fn);
```
Actually could we update the error message to be even more precise on this case: you did a differentiable view on a view created in a no_grad block and are now modifying inplace.

soulitzer replied:
Good idea, updated.

albanD reviewed (commented):
Nice catch!

soulitzer committed:
Update on "Remove wrong internal assert in handle_view_on_rebase"


Fixes: https://github.com/pytorch/pytorch/issues/88205

The `CreationMeta::NO_GRAD_MODE` path in handle_view_on_rebase wrongly assumes that the tensor would be a leaf, because tensors created in no_grad are always leaf tensors. However, due to creation_meta propagation, a view of a view created in no_grad also has `CreationMeta::NO_GRAD_MODE`, but DOES have grad_fn.

[ghstack-poisoned]

soulitzer committed:
Update on "Remove wrong internal assert in handle_view_on_rebase"


Fixes: https://github.com/pytorch/pytorch/issues/88205

The `CreationMeta::NO_GRAD_MODE` path in handle_view_on_rebase wrongly assumes that the tensor would be a leaf, because tensors created in no_grad are always leaf tensors. However, due to creation_meta propagation, a view of a view created in no_grad also has `CreationMeta::NO_GRAD_MODE`, but DOES have grad_fn.

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -f "Unrelated failures"
